src/impl_constructors.rs

```diff
@@ -275,51 +289,52 @@ impl<S, A, D> ArrayBase<S, D>
         if shape.custom {
-            Self::from_vec_dim_stride(shape.dim, shape.strides, v)
+            dimension::can_index_slice(&v, &dim, &strides)?;
         } else {
-            let dim = shape.dim;
-            let strides = shape.strides;
-            if dim.size_checked() != Some(v.len()) {
+            dimension::can_index_slice_not_custom::<A, _>(&v, &dim)?;
+            if dim.size() != v.len() {
                 return Err(error::incompatible_shapes(&Ix1(v.len()), &dim));
             }
-            unsafe { Ok(Self::from_vec_dim_stride_unchecked(dim, strides, v)) }
         }
+        unsafe { Ok(Self::from_vec_dim_stride_unchecked(dim, strides, v)) }
     }
 
-    /// Create an array from a vector and interpret it according to the
-    /// provided dimensions and strides. (No cloning of elements needed.)
+    /// Creates an array from a vector and interpret it according to the
+    /// provided shape and strides. (No cloning of elements needed.)
+    ///
+    /// The caller must ensure that the following conditions are met:
+    ///
+    /// 1. The ndim of `dim` and `strides` must be the same.
+    ///
+    /// 2. The product of non-zero axis lengths must not exceed `isize::MAX`.
+    ///
+    /// 3. For axes with length > 1, the stride must be nonnegative.
     ///
-    /// Unsafe because dimension and strides are unchecked.
+    /// 4. If the array will be empty (any axes are zero-length), the
+    ///    difference between the least address and greatest address accessible
+    ///    by moving along all axes must be ≤ `v.len()`.
+    ///
+    ///    If the array will not be empty, the difference between the least
+    ///    address and greatest address accessible by moving along all axes
+    ///    must be < `v.len()`.
+    ///
+    /// 5. The strides must not allow any element to be referenced by two different
+    ///    indices.
     pub unsafe fn from_shape_vec_unchecked<Sh>(shape: Sh, v: Vec<A>) -> Self
         where Sh: Into<StrideShape<D>>,
     {
         let shape = shape.into();
         Self::from_vec_dim_stride_unchecked(shape.dim, shape.strides, v)
     }
 
-    fn from_vec_dim_stride(dim: D, strides: D, v: Vec<A>)
-        -> Result<Self, ShapeError>
-    {
-        dimension::can_index_slice(&v, &dim, &strides).map(|_| {
-            unsafe {
-                Self::from_vec_dim_stride_unchecked(dim, strides, v)
-            }
-        })
-    }
-
     unsafe fn from_vec_dim_stride_unchecked(dim: D, strides: D, mut v: Vec<A>)
         -> Self
     {
         // debug check for issues that indicates wrong use of this constructor
-        debug_assert!(match dimension::can_index_slice(&v, &dim, &strides) {
-            Ok(_) => true,
-            Err(ref e) => match e.kind() {
-                ErrorKind::OutOfBounds => false,
-                ErrorKind::RangeLimited => false,
-                _ => true,
-            }
-        });
+        debug_assert!(dimension::can_index_slice(&v, &dim, &strides).is_ok());
```

Comment:
Is this now correct? There were some negative stride issues we couldn't check properly before.

Reply:
See my comment below. If we want to allow negative strides, we need to provide some way for the user to specify the `ptr` since it's always different from `vec.as_mut_ptr()` when there are negative strides for axes with length > 1.